base/compiler/typeutils.jl

```diff
@@ -133,10 +133,16 @@ function switchtupleunion(@nospecialize(ty))
     return _switchtupleunion(Any[tparams...], length(tparams), [], ty)
 end
 
+switchtupleunion(t::Vector{Any}) = _switchtupleunion(t, length(t), [], nothing)
+
 function _switchtupleunion(t::Vector{Any}, i::Int, tunion::Vector{Any}, @nospecialize(origt))
     if i == 0
-        tpl = rewrap_unionall(Tuple{t...}, origt)
-        push!(tunion, tpl)
+        if origt === nothing
+            push!(tunion, t)
```

Comment:
```suggestion
            push!(tunion, copy(t))
```

I found we need to copy `t` here, otherwise it will be mutated at `t[i] = ti` in recursion.